### ISSUE TITLE
graceful shutdown only fires for `IOException`s

### ISSUE THREAD

**Issue by mitchellwrosen** [yesodweb/wai#708]
Hi, I discovered that the graceful shutdown mechanism only works when the web server experiences an `IOException`. Is this the intended behavior? (It makes it so Ctrl-C hard-quits a warp server without gracefully shutting down, for example) Here is the relevant code from `Network.Wai.Handler.Warp.Run`:

```haskell
acceptConnection :: Settings
                 -> IO (IO (Connection, Transport), SockAddr)
                 -> Application
                 -> Counter
                 -> InternalInfo0
                 -> IO ()
acceptConnection set getConnMaker app counter ii0 = do
    void $ mask_ acceptLoop
    gracefulShutdown set counter
  where
    acceptLoop = do
        allowInterrupt

        mx <- acceptNewConnection
        case mx of
            Nothing             -> return ()
            Just (mkConn, addr) -> do
                fork set mkConn addr app counter ii0
                acceptLoop

    acceptNewConnection = do
        ex <- try getConnMaker
        case ex of
            Right x -> return $ Just x
            Left e -> do
                let eConnAborted = getErrno eCONNABORTED
                    getErrno (Errno cInt) = cInt
                if ioe_errno e == Just eConnAborted
                    then acceptNewConnection
                    else do
                        settingsOnException set Nothing $ toException e
                        return Nothing
```

The `try getConnMaker` is called at type `IO a -> IO (Either IOException a)`.

Thanks.

**Comment by snoyberg** [comment]
This code isn't for graceful shutdown. It's to deal with a specific corner case of file descriptors becoming exhausted IIRC (CC @kazu-yamamoto). We should probably have a comment explaining that.

We could change this by using `finally` or similar to ensure that graceful shutdown kicks in, but I'm not sure it will behave properly in cases of things like async exceptions. The real question is what the expected semantics of this feature are in the different kinds of shutdown scenarios, which I'm not sure of.

**Comment by mitchellwrosen** [comment]
Can you clarify what you mean by this?

> This code isn't for graceful shutdown.

Is there some other code that is for graceful shutdown?

The basic problem I'm trying to solve: Ctrl-C on a Warp server only shuts down after all outstanding requests are handled. Is that possible with the current Warp API? Thanks :)

**Comment by snoyberg** [comment]
What I meant is that the error checking code you pointed at is not intended for graceful shutdown, it's dealing with a specific corner case. In other words, changing that bit of code would not be appropriate. That's why I mentioned the potential of using `finally`.

> The basic problem I'm trying to solve: Ctrl-C on a Warp server only shuts down after all outstanding requests are handled. Is that possible with the current Warp API? Thanks :)

Can you rephrase this as "I expected X but Y happened?" I'm not sure what you're saying is currently happening versus what you want to happen.

**Comment by mitchellwrosen** [comment]
I want to be able to, upon receiving an OS signal, reject all incoming requests and wait for all outstanding requests to finish. Warp already counts outstanding requests in a TVar and exposes a "graceful shutdown" mechanism, but I am unsure of how to use it to get this behavior.

**Comment by snoyberg** [comment]
Changing the code to use `finally` should address that goal. My concern is that this almost the antithesis of good async exception handling behavior: instead of cleaning up quickly and exiting, it's going to delay an unbounded amount of time. I'm OK with that as opt-in behavior if it's well documented. It seems like the "graceful shutdown" mechanism is already opt-in enough that making this change to it wouldn't be breaking for anyone.

Does this make sense?

**Comment by kazu-yamamoto** [comment]
> I want to be able to, upon receiving an OS signal, reject all incoming requests and wait for all outstanding requests to finish. Warp already counts outstanding requests in a TVar and exposes a "graceful shutdown" mechanism, but I am unsure of how to use it to get this behavior.

Sorry for the delay. I was on vacation.

`mighttpd2` uses Warp's graceful shutdown. `mighttpd2` sets a signal handler which closes the *listening* socket. This breaks `acceptLoop`, so `gracefulShutdown` is executed.

I agree that
- we should use `finally` to ensure that `gracefulShutdown` is called
- we should prepare a new API to break `acceptLoop` (such as `GracefulTeardown`)

**Comment by mitchellwrosen** [comment]
@kazu-yamamoto Thanks! Is there anything more you would like me to do to help hash out the implementation? I am a bit confused by your two bullet points - if we want to always enter `gracefulShutdown` (as `finally` does not distinguish exception types), then why is there a need for a new `GracefulTeardown` code path in the accept loop?

**Comment by kazu-yamamoto** [comment]
Never mind. After investigations, I changed my opinion. The current code should be kept as it is.

Typically I run `mighty` in two mode:
- Daemon: I typically send a signal to let `mighty` enter graceful shutdown and executes another `mighty`.
- Debug: `mighty` is not detached from its terminal. I use "C-c" to stop `mighty`. In this mode, graceful shutdown is not necessary.

If we use `finally`, graceful shutdown is always called. This is very inconvenient for debug mode since I have to wait a few minutes until the graceful shutdown finishes.

I added some comments to explain this.

**Comment by kazu-yamamoto** [comment]
@mitchellwrosen So, my question: Is closing a listening socket good enough for your purpose? Or are any other APIs for graceful shutdown necessary?

**Comment by mitchellwrosen** [comment]
@kazu-yamamoto 

Here is the `gracefulShutdown` code from warp:

```
gracefulShutdown :: Settings -> Counter -> IO ()
gracefulShutdown set counter =
    case settingsGracefulShutdownTimeout set of
        Nothing ->
            waitForZero counter
        (Just seconds) ->
            void (timeout (seconds * microsPerSecond) (waitForZero counter))
            where microsPerSecond = 1000000
```

This is because warp is already counting requests, so it already knows how many requests are outstanding, and (to my knowledge) does not expose this count in the external API.

Can you clarify how you accomplish this:

> Daemon: I typically send a signal to let mighty enter graceful shutdown and executes another mighty.

Are you counting requests outside of warp, and _not_ using its built-in graceful shutdown mechanism?

**Comment by mitchellwrosen** [comment]
In the current implementation, I can get the behavior I want only by throwing an `IOException` to the thread running warp. Any other exception does not cause `gracefulShutdown` to be called. That was the original reason I opened this ticket - to clarify that this was the intended behavior.

**Comment by kazu-yamamoto** [comment]
> Are you counting requests outside of warp, and not using its built-in graceful shutdown mechanism?

No. `mighty` uses the built-in graceful shutdown. `mighty` prepares a signal handler which closes a listening socket. So, when I send the signal to `mighty`, the build-in graceful shutdown runs.

**Comment by kazu-yamamoto** [comment]
> In the current implementation, I can get the behavior I want only by throwing an IOException to the thread running warp. Any other exception does not cause gracefulShutdown to be called. That was the original reason I opened this ticket - to clarify that this was the intended behavior.

Ah. OK. I understand your point.

I don't think that this behavior is intentional. The current code satisfies my requirement. But if it does not satisfy your requirement, you can change it as you like. Would you send us a PR?

**Comment by mitchellwrosen** [comment]
> No. mighty uses the built-in graceful shutdown. mighty prepares a signal handler which closes a listening socket. So, when I send the signal to mighty, the build-in graceful shutdown runs.

Oh, I see. You are inducing an `IOException` by closing the listening socket! This is in fact explain in the documentation of `runSettingsSocket`:

> When the listen socket in the second argument is closed, all live connections are gracefully shut down.

Thanks, I will use this technique as well. All's good, shall I close this issue?

**Comment by mitchellwrosen** [comment]
Sure, I'll make a PR with some documentation improvements

**Comment by kazu-yamamoto** [comment]
@mitchellwrosen ping

**Comment by mitchellwrosen** [comment]
@kazu-yamamoto #716 Pong :)

**Comment by kazu-yamamoto** [comment]
Thanks. Let's close this.
